### PR TITLE
Migrate login page to Spark/React

### PR DESCRIPTION
**Pull request by lneves12**
## What does this PR change?

Migration of the login page to react+spark.

**Highlights:**

- New template for login page. No more template hacks + reduced page size
- Introducing react-hooks

Reviewers: best reviewed on a commit-to-commit basis.

## GUI diff

image

- [X] **DONE**

## Documentation
- No documentation needed: the same functionality as before

- [X] **DONE**

## Test coverage
- Unit tests were added
- Cucumber tests were adjusted

- [X] **DONE**

## Links

Fixes https://github.com/SUSE/spacewalk/issues/5556

- [X] **DONE**

## Changelogs

Copy the following sentence as a new comment if you don't need changelog entries:

> gitarro no changelog needed !!!

If the test `changelog_test`already run, then add another new comment with the following text:

> gitarro rerun changelog_test !!!

### CODE REVIEW

**Comment by paususe** [comment]
The more we wait on this PR, the more code that gets broken. Can we please fix the issues and merge?

**Comment by mbologna** [comment]
> The more we wait on this PR, the more code that gets broken. Can we please fix the issues and merge?

We will dedicate part of the next sprint to make it ready for merging.

**Comment by mbologna** [comment]
Note to self:
- [ ] JUnit Tests
- [ ] Cucumber TEST project
- [ ] Migrate old setup to new setup: check Apache config overwrite
- [ ] Merging :rocket:

**Comment by mbologna** [comment]
Testsuite: setup done, fixed 2 bugs, testsuite runs in progress. Not yet satisfied with the failures.

**Comment by mbologna** [comment]
Apache `%config` check:

```
/etc/apache2/conf.d # cp zz-spacewalk-www.conf /root/
# zypper in --oldpackage ./spacewalk-config-4.0.10-400.2.1.uyunimaster.noarch.rpm 
/etc/apache2/conf.d # diff -u zz-spacewalk-www.conf /root/zz-spacewalk-www.conf 
--- zz-spacewalk-www.conf	2019-08-20 18:47:26.000000000 +0200
+++ /root/zz-spacewalk-www.conf	2019-08-21 10:42:40.828695971 +0200
@@ -43,7 +43,7 @@
 RewriteRule ^/rpc/api /rhn/rpc/api
 RewriteRule ^/ks/dist(.*)$ /rhn/common/DownloadFile.do?url=/ks/dist$1
 RewriteRule ^(/ty/.*)$ /rhn/common/DownloadFile.do?url=$1
-RewriteRule ^/index\.html$ /rhn/manager/login
+RewriteRule ^/index\.html$ /rhn/Login.do
 
 # For rhn-custom-info
 RewriteRule ^/WEBRPC /rhn/rpc/api
@@ -63,8 +63,7 @@
 <Proxy ajp://localhost:8009>
   ProxySet min=1
 </Proxy>
-RewriteRule ^/rhn/Login2\.do ajp://localhost:8009/rhn/manager/login [P]
-RewriteRule ^/rhn/(Re)?Login(.*)$ /rhn/manager/login
+RewriteRule ^/rhn/Login2\.do ajp://localhost:8009/rhn/Login.do [P]
 RewriteCond %{REQUEST_URI} !^/rhn/websocket
 RewriteRule ^/rhn(.*) ajp://localhost:8009/rhn$1 [P]
 </IfModule>
```

OK.

File: java/code/webapp/WEB-INF/decorators/layout_head.jsp [hunk 1]
```diff
@@ -49,10 +49,12 @@
     <script src="/javascript/loggerhead.js?cb=${cb_version}"></script>
     <script src="/javascript/frontend-log.js?cb=${cb_version}"></script>
 
+
```
Comment (renner) [inline java/code/webapp/WEB-INF/decorators/layout_head.jsp]:
```suggestion
```

File: java/code/src/com/redhat/rhn/frontend/security/PxtAuthenticationService.java [hunk 2]
```diff
@@ -48,6 +47,7 @@
         set.add("/rhn/Login");
         set.add("/rhn/ReLogin");
```
Comment (renner) [inline java/code/src/com/redhat/rhn/frontend/security/PxtAuthenticationService.java]: Those two we URLs were removed from `struts-config.xml` below, so it might be safe to remove them from this set as well?

File: spacewalk/config/etc/httpd/conf.d/zz-spacewalk-www.conf [hunk 3]
```diff
@@ -63,7 +63,8 @@ ProxyTimeout 600
 <Proxy ajp://localhost:8009>
   ProxySet min=1
 </Proxy>
-RewriteRule ^/rhn/Login2\.do ajp://localhost:8009/rhn/Login.do [P]
+RewriteRule ^/rhn/Login2\.do ajp://localhost:8009/rhn/manager/login [P]
+RewriteRule ^/rhn/(Re)?Login(.*)$ /rhn/manager/login
```
Comment (renner) [inline spacewalk/config/etc/httpd/conf.d/zz-spacewalk-www.conf]: Do we still need to keep `Login2.do` and this redirect around? And why do we need a redirect from `Login` / `ReLogin` here?

File: susemanager-frontend/package.json [hunk 4]
```diff
@@ -1,8 +1,13 @@
...
   "workspaces": {
-    "packages": ["susemanager-nodejs-sdk-devel", "../web/html/src"],
-    "nohoist": ["**"]
+    "packages": [
+      "susemanager-nodejs-sdk-devel",
+      "../web/html/src"
+    ],
+    "nohoist": [
+      "**"
+    ]
```
Comment (renner) [inline susemanager-frontend/package.json]: Is this change relevant to the patch (it doesn't seem to be)?

File: java/code/src/com/redhat/rhn/frontend/struts/test/RequestContextTest.java [hunk 5]
```diff
@@ -38,12 +35,12 @@ public RequestContextTest(String name) {
...
-        HttpServletRequest request = ltest.loginUserIntoSessionTest();
-        RequestContext requestContext = new RequestContext(request);
-        assertNotNull(requestContext.getCurrentUser());
-    }
+//    public final void testGetLoggedInUser() throws Exception {
+//        LoginActionTest ltest = new LoginActionTest();
+//        HttpServletRequest request = ltest.loginUserIntoSessionTest();
+//        RequestContext requestContext = new RequestContext(request);
+//        assertNotNull(requestContext.getCurrentUser());
+//    }
```
Comment (renner) [inline java/code/src/com/redhat/rhn/frontend/struts/test/RequestContextTest.java]: This commented out test seems to be equal to the previous one, remove it?

File: java/code/src/com/suse/manager/webui/controllers/test/LoginControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+//                new ActionForward(RhnHelper.DEFAULT_FORWARD, "path", false));
+//        RhnMockDynaActionForm form = new RhnMockDynaActionForm("loginForm");
+//        RhnMockHttpServletRequest req = new RhnMockHttpServletRequest();
+//        RhnMockHttpServletResponse resp = new RhnMockHttpServletResponse();
+//        req.setSession(new MockHttpSession());
+//        req.setupServerName("mymachine.rhndev.redhat.com");
+//
+//        ActionForward rc = action.execute(mapping, form, req, resp);
+//        assertNotNull(rc);
+//    }
```
Comment (renner) [inline java/code/src/com/suse/manager/webui/controllers/test/LoginControllerTest.java]: Let's not merge with commented out tests, can we update those tests to the new implementation?

File: java/code/src/com/redhat/rhn/frontend/struts/test/RequestContextTest.java (hunk 5 above)
Comment (mbologna) [reply java/code/src/com/redhat/rhn/frontend/struts/test/RequestContextTest.java]: Test adapted

File: java/code/src/com/suse/manager/webui/controllers/test/LoginControllerTest.java (hunk 6 above)
Comment (mbologna) [reply java/code/src/com/suse/manager/webui/controllers/test/LoginControllerTest.java]: Tests has been finally written!

File: java/code/src/com/redhat/rhn/frontend/security/PxtAuthenticationService.java (hunk 2 above)
Comment (mbologna) [reply java/code/src/com/redhat/rhn/frontend/security/PxtAuthenticationService.java]: We removed them from struts but we enabled via Apache Rewrite a redirect of those URLs to the new login page, just for safekeeping.

File: spacewalk/config/etc/httpd/conf.d/zz-spacewalk-www.conf (hunk 3 above)
Comment (mbologna) [reply spacewalk/config/etc/httpd/conf.d/zz-spacewalk-www.conf]: AFAIK Login2.do is used for Kerberos authentication. The Login/Relogin has been added as per the comment above.

File: susemanager-frontend/package.json (hunk 4 above)
Comment (mbologna) [reply susemanager-frontend/package.json]: @lneves12 what's your take on this?

Comment (lneves12) [reply susemanager-frontend/package.json]: Addressed :)

**Comment by lneves12** [comment]
I added some more commits cleaning up some code and adding the loginLength/passwordLength properties. This is starting to look good to merge. The last testsuite results also looked promising

**Comment by mbologna** [comment]
Added screenshot and information on the PR opening text

**Comment by mbologna** [comment]
I squashed all the commits for an easy review.
Testsuite results: I pushed an additional fix.
All the other testsuite failures seem not related to the login page.

Reviewers, please do your game. As soon as we have consensus, this is ready to be merged IMHO.

**Comment by mbologna** [comment]
Adding @mantel as a reviewer just for spacewalk/config/etc/httpd/conf.d/zz-spacewalk-www.conf: Hubert, we are changing `/etc/apache2/conf.d/zz-spacewalk-www.conf` file in this PR, which is part of the `spacewalk-config` RPM package.
If [I did my homework properly](https://github.com/uyuni-project/uyuni/pull/564#issuecomment-523361018) when we upgrade this package the new files (the ones that come with the package) will overwrite the old files (the ones that are in the system) [also according to https://www.cl.cam.ac.uk/~jw35/docs/rpm_config.html].
Am I missing anything?

**Comment by mbologna** [comment]
For reference: last run with the tip of this branch development version:

22 failed, 0 pending, 100 skipped, 0 undefined, 891 passed

https://ci.suse.de/job/manager-TEST-Orion-cucumber/ws/results/build-356/output.html

**Comment by mantel** [comment]
> Adding @mantel as a reviewer just for spacewalk/config/etc/httpd/conf.d/zz-spacewalk-www.conf: Hubert, we are changing `/etc/apache2/conf.d/zz-spacewalk-www.conf` file in this PR, which is part of the `spacewalk-config` RPM package.
> If [I did my homework properly](https://github.com/uyuni-project/uyuni/pull/564#issuecomment-523361018) when we upgrade this package the new files (the ones that come with the package) will overwrite the old files (the ones that are in the system) also according to https://www.cl.cam.ac.uk/~jw35/docs/rpm_config.html.
> Am I missing anything?

If the file is tagged as %config then it should be fine. I assume you already tested the actual change :)

Sorry for missing this review request; just noticed it is already one complete week old :/

**Comment by mbologna** [comment]
Target changed to Uyuni ONLY

File: java/code/src/com/redhat/rhn/frontend/action/LoginHelper.java [hunk 7]
```diff
@@ -261,9 +255,8 @@ private static String decodeFromIso88591(String string, String defaultString) {
      * @param request actual request
      * @param response actual reponse
      * @param user logged in user
-     * @return returns true, if redirect
-     */
-    public static boolean successfulLogin(HttpServletRequest request,
+     * */
```
Comment (renner) [inline java/code/src/com/redhat/rhn/frontend/action/LoginHelper.java]: This javadoc comment ending was probably a typo.

File: java/code/src/com/redhat/rhn/frontend/action/LoginHelper.java [hunk 8]
```diff
@@ -273,26 +266,38 @@ public static boolean successfulLogin(HttpServletRequest request,
...
-        catch (IOException e) {
-            e.printStackTrace();
+        if (requestMethod != null && requestMethod.equals("POST")) {
+            urlBounce = DEFAULT_URL_BOUNCE;
         }
-        return false;
+        return urlBounce;
     }
 
+
```
Comment (renner) [inline java/code/src/com/redhat/rhn/frontend/action/LoginHelper.java]:
```suggestion
```

File: java/code/src/com/redhat/rhn/frontend/security/PxtAuthenticationService.java (hunk 2 above)
Comment (renner) [reply java/code/src/com/redhat/rhn/frontend/security/PxtAuthenticationService.java]: Okay, but why would that new redirect be needed exactly? If someone has a bookmark to those URLs they should update it, it's okay from my perspective that an old bookmark stops working. If there is broken links from inside the application those should be fixed, and with a redirect in place we might never find out if there is a broken link somewhere. Am I missing some other usecase?

File: java/code/src/com/redhat/rhn/frontend/servlets/CreateRedirectURI.java [hunk 9]
```diff
@@ -14,13 +14,12 @@
  */
 package com.redhat.rhn.frontend.servlets;
 
-import com.redhat.rhn.frontend.action.LoginAction;
+import com.redhat.rhn.frontend.action.LoginHelper;
```
Comment (renner) [inline java/code/src/com/redhat/rhn/frontend/servlets/CreateRedirectURI.java]: If `LoginAction` has turned into a `LoginHelper` class, is the `com.redhat.rhn.frontend.action` package still the appropriate place for this class?

File: java/code/src/com/suse/manager/webui/controllers/login/LoginController.java [hunk 10]
```diff
@@ -0,0 +1,228 @@
+/**
+ * Copyright (c) 2018 SUSE LLC
```
Comment (renner) [inline java/code/src/com/suse/manager/webui/controllers/login/LoginController.java]: If this is a new class you want to put 2019 here.

File: spacewalk/config/etc/httpd/conf.d/zz-spacewalk-www.conf (hunk 3 above)
Comment (renner) [reply spacewalk/config/etc/httpd/conf.d/zz-spacewalk-www.conf]: As mentioned above, I would still question the need of the new redirect (`Login/ReLogin`), but also the existence of `Login2.do` in general. I would be okay to keep it if it is needed for Kerberos authentication, but I would prefer if we properly understand the reason.

File: java/code/src/com/redhat/rhn/frontend/action/LoginHelper.java [hunk 11]
```diff
@@ -44,23 +44,18 @@
...
-import java.io.IOException;
 import java.io.UnsupportedEncodingException;
 import java.util.ArrayList;
 import java.util.Date;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
-
+import javax.security.auth.login.LoginException;
```
Comment (renner) [inline java/code/src/com/redhat/rhn/frontend/action/LoginHelper.java]: This does not matter so much and you can decide to ignore it, but the original [upstream import order configuration](https://github.com/uyuni-project/uyuni/blob/master/java/conf/eclipse/rhn.importorder) would keep `java` imports separately from `javax`.

Comment (mbologna) [reply java/code/src/com/redhat/rhn/frontend/action/LoginHelper.java]: OK, I updated the IntelliJ wiki rules to reflect this: https://github.com/uyuni-project/uyuni/wiki/IntelliJ-IDEA-specific-development-instructions/_compare/8da7f60de53dcd4f17b9428e3ced6ec8fe468fdb

File: java/code/src/com/redhat/rhn/frontend/security/PxtAuthenticationService.java (hunk 2 above)
Comment (mbologna) [reply java/code/src/com/redhat/rhn/frontend/security/PxtAuthenticationService.java]: OK, removed ReLogin and made it point to the new login page.
Login2 must stay instead: it is used for IPA authentication: the first login fails (expected, it checks against Kerberos) and we require another authentication (using local authentication).

File: java/code/src/com/redhat/rhn/frontend/servlets/CreateRedirectURI.java (hunk 9 above)
Comment (mbologna) [reply java/code/src/com/redhat/rhn/frontend/servlets/CreateRedirectURI.java]: Moved into another package

File: java/code/src/com/suse/manager/webui/controllers/login/LoginController.java (hunk 10 above)
Comment (mbologna) [reply java/code/src/com/suse/manager/webui/controllers/login/LoginController.java]: Fixed

File: spacewalk/config/etc/httpd/conf.d/zz-spacewalk-www.conf (hunk 3 above)
Comment (mbologna) [reply spacewalk/config/etc/httpd/conf.d/zz-spacewalk-www.conf]: Ok, removed as per the comment above

**Comment by mbologna** [comment]
@renner we addressed all your requests, would you mind taking another review, please?

File: testsuite/features/secondary/srv_menu.feature [hunk 12]
```diff
@@ -116,7 +116,7 @@ Feature: Web UI - Main landing page menu, texts and links
 
   Scenario: The manager configuration submenu of the admin menu
     When I follow the left menu "Admin > Manager Configuration"
-    Then I should see a "SUSE Manager Configuration - General Configuration" text
+    Then I should see a "Configuration - General Configuration" text
```
Comment (renner) [inline testsuite/features/secondary/srv_menu.feature]: This change seems unrelated, maybe it has been changed in master meanwhile?

File: java/conf/audit/auditlog-config.yaml [hunk 13]
```diff
@@ -1235,4 +1231,3 @@
 
 /network/users/details/contact_methods/edit.pxt:
     type: ACTION_CHANGE
-
```
Comment (renner) [inline java/conf/audit/auditlog-config.yaml]: Unrelated removal of newline, maybe this was done by your editor?

File: proxy/proxy/httpd-conf/spacewalk-proxy.conf [hunk 14]
```diff
@@ -52,13 +52,11 @@
...
-   RewriteRule ^/rhn/?$ https://%{SERVER_NAME}/rhn/Login.do  [R,L]
+   RewriteRule ^/rhn/?$ https://%{SERVER_NAME}/rhn/manager/login  [R,L]
 </IfModule>
 
 
 
 #Disable SSL2, left only higher
 SSLProtocol all -SSLv2
 SSLProxyEngine on
-
```
Comment (renner) [inline proxy/proxy/httpd-conf/spacewalk-proxy.conf]: Unrelated removal of newline.

File: testsuite/features/secondary/srv_menu.feature (hunk 12 above)
Comment (mbologna) [reply testsuite/features/secondary/srv_menu.feature]: This is a bugfix: for Uyuni the text shown will be `Uyuni Configuration` - this change will fix the failing testsuite test

File: proxy/proxy/httpd-conf/spacewalk-proxy.conf (hunk 14 above)
Comment (mbologna) [reply proxy/proxy/httpd-conf/spacewalk-proxy.conf]: Fixed

File: web/spacewalk-web.changes [hunk 15]
```diff
@@ -827,7 +828,7 @@ Thu Feb 13 15:12:54 CET 2014 - mc@suse.de
 Fri Feb 07 12:58:53 CET 2014 - mc@suse.de
 
 - version 2.1.54.1-1
-- changes in the logic to update the tick icon. 
+- changes in the logic to update the tick icon.
```
Comment (mcalmer) [inline web/spacewalk-web.changes]: No whitespace cleanup in changes files please.

**Comment by mbologna** [comment]
gitarro rerun java_pgsql_tests !!!
